src/ast/recfun_decl_plugin.h

```diff
@@ -0,0 +1,264 @@
+/*++
+Module Name:
+
+    recfun_decl_plugin.h
+
+Abstract:
+
+    Declaration and definition of (potentially recursive) functions
+
+Author:
+
+    Simon Cruanes 2017-11
+
+Revision History:
+
+--*/
+
+#pragma once
+
+#include "ast/ast.h"
+#include "ast/rewriter/th_rewriter.h"
+
+namespace recfun {
+    class case_def; //<! one possible control path of a function
+    class case_pred; //<! a predicate guarding a given control flow path of a function
+    class util; //<! util for other modules
+    class def; //!< definition of a (recursive) function
+
+    enum op_kind {
+        OP_FUN_MACRO, // unfold eagerly, only one case
+        OP_FUN_DEFINED, // defined function with multiple cases, possibly recursive
+        OP_FUN_CASE_PRED, // predicate guarding a given control flow path
+    };
+
+    /*! A predicate `p(t1…tn)`, that, if true, means `f(t1…tn)` is following
+        a given path of its control flow and can be unrolled.
+
+        For example, `fact n := if n<2 then 1 else n * fact(n-1)` would have two cases,
+        and therefore two case predicates `C_fact_0` and `C_fact_1`, where
+        `C_fact_0(t)=true` means `t<2` (first path) and `C_fact_1(t)=true` means `¬(t<2)` (second path).
+    */
+    class case_pred {
+        friend class case_def;
+        symbol                      m_name; //<! symbol for the predicate
+        std::string                 m_name_buf; //<! memory for m_name
+        sort_ref_vector const &     m_arg_sorts;
+
+        case_pred(sort_ref_vector const & args) : m_name(), m_name_buf(), m_arg_sorts(args) {}
+        void set_name(std::string & s);
+    public:
+        symbol const & get_name() const { return m_name; }
+        sort_ref_vector const & get_args() const { return m_arg_sorts; }
+        unsigned get_num_args() const { return m_arg_sorts.size(); }
+    };
+
+    typedef ptr_vector<var>  vars;
+
+    class case_def {
+        friend class def;
+        case_pred           m_pred; //<! predicate used for this case
+        vector<expr_ref>    m_guards; //<! conjunction that is equivalent to this case
```

Comment:
expr_ref_vector?
I know it is actually not as convenient as vector<expr_ref> when accessing elements.